src/sortTemplate.ts

```diff
@@ -0,0 +1,297 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as vscode from "vscode";
+import { Language } from "../extension.bundle";
+import { templateKeys } from './constants';
+import { DeploymentTemplate } from "./DeploymentTemplate";
+import { ext } from './extensionVariables';
+import { IParameterDefinition } from './IParameterDefinition';
+import * as Json from "./JSON";
+import * as language from "./Language";
+import { UserFunctionDefinition } from './UserFunctionDefinition';
+import { UserFunctionNamespaceDefinition } from './UserFunctionNamespaceDefinition';
+import { IVariableDefinition } from './VariableDefinition';
+
+export enum SortType {
+    Resources,
+    Outputs,
+    Parameters,
+    Variables,
+    Functions,
+    TopLevel
+}
+
+export class SortQuickPickItem implements vscode.QuickPickItem {
+    public label: string;
+    public value: SortType;
+    public description: string;
+
+    constructor(label: string, value: SortType, description: string) {
+        this.label = label;
+        this.value = value;
+        this.description = description;
+    }
+}
+
+export function getQuickPickItems(): SortQuickPickItem[] {
+    let items: SortQuickPickItem[] = [];
+    items.push(new SortQuickPickItem("Functions", SortType.Functions, "Sort function namespaces and functions"));
+    items.push(new SortQuickPickItem("Outputs", SortType.Outputs, "Sort outputs"));
+    items.push(new SortQuickPickItem("Parameters", SortType.Parameters, "Sort parameters for the template"));
+    items.push(new SortQuickPickItem("Resources", SortType.Resources, "Sort resources based on the name including first level of child resources"));
+    items.push(new SortQuickPickItem("Variables", SortType.Variables, "Sort variables"));
+    items.push(new SortQuickPickItem("Top level", SortType.TopLevel, "Sort top level items based on recommended order (parameters, functions, variables, resources, outputs)"));
+    return items;
```

Comment:
What do you think about adding an "All"?

Reply:
I interpreted the feedback from @bmoore-msft as it is preferable to only sort small parts that are clearly specified by the user (sort variables), then the user can check the variables and see if the sorting generated the expected result.